compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -2168,14 +2208,19 @@ public void visit(BLangIndexBasedAccess indexBasedAccessExpr) {
 
         BType exprType = indexBasedAccessExpr.expr.type;
         BLangExpression indexExpr = indexBasedAccessExpr.indexExpr;
+
         if (actualType != symTable.semanticError &&
-                types.isSubTypeOfBaseType(exprType, TypeTags.RECORD) &&
-                (indexBasedAccessExpr.lhsVar || indexBasedAccessExpr.compoundAssignmentLhsVar) &&
-                (indexExpr.getKind() == NodeKind.LITERAL || isConst(indexExpr))) {
-            String name = getConstFieldName(indexExpr);
-            if (allRecordsHaveRequiredReadonlyField(exprType, name)) {
+                (indexBasedAccessExpr.lhsVar || indexBasedAccessExpr.compoundAssignmentLhsVar)) {
+            if (isAllReadonlyTypes(exprType)) {
```

Comment:
Check for optional fields.

Reply:
This is not required since what we are checking here is something like

```
(Foo & readonly)|(Bar & readonly) x = {};
x["val"] = 1;
```

This is always invalid since the value itself is immutable and cannot be updated.